h5py/_npystrings.pyx

```diff
@@ -0,0 +1,255 @@
+    cdef extern from "numpy/ndarraytypes.h":
+        ctypedef struct npy_string_allocator:
+            pass
+
+        ctypedef struct npy_packed_static_string:
+            pass
+
+        ctypedef struct npy_static_string:
+            size_t size
+            const char *buf
+
+        ctypedef struct PyArray_StringDTypeObject:
+            PyArray_Descr base
+            PyObject *na_object
+            char coerce
+            char has_nan_na
+            char has_string_na
+            char array_owned
+            npy_static_string default_string
+            npy_static_string na_name
+            npy_string_allocator *allocator
+
+    cdef extern from "numpy/arrayobject.h":
+        npy_string_allocator *NpyString_acquire_allocator(const PyArray_StringDTypeObject *descr)
+        void NpyString_acquire_allocators(size_t n_descriptors, PyArray_Descr *const descrs[], npy_string_allocator *allocators[])
+        void NpyString_release_allocator(npy_string_allocator *allocator)
+        void NpyString_release_allocators(size_t length, npy_string_allocator *allocators[])
+        int NpyString_load(npy_string_allocator *allocator, const npy_packed_static_string *packed_string, npy_static_string *unpacked_string)
+        int NpyString_pack_null(npy_string_allocator *allocator, npy_packed_static_string *packed_string)
+        int NpyString_pack(npy_string_allocator *allocator, npy_packed_static_string *packed_string, const char *buf, size_t size)
+
+ELSE:
+    from numpy cimport (
+        npy_string_allocator,
+        npy_static_string,
+        npy_packed_static_string,
+        NpyString_pack,
+        NpyString_load,
+        NpyString_acquire_allocator,
+        NpyString_release_allocator,
+        PyArray_StringDTypeObject,
+    )
+
+# Can't call sizeof(npy_packed_static_string) because the struct is
+# internal to numpy
+cdef size_t SIZEOF_NPY_PACKED_STATIC_STRING = np.dtype("T").itemsize
+
+ctypedef struct vstrings_scatter_t:
+    size_t i
+    npy_string_allocator *allocator
+    const char **contig
+
+ctypedef struct vstrings_gather_t:
+    size_t i
+    npy_string_allocator *allocator
+    npy_static_string *unpacked
+
+
+cdef herr_t vstrings_scatter_cb(
+    void* elem, hid_t type_id, unsigned ndim,
+    const hsize_t *point, void *operator_data
+) except -1:
+    cdef vstrings_scatter_t* info = <vstrings_scatter_t*>operator_data
+    cdef const char* buf = info[0].contig[info[0].i]
+    # Deep copy char* from h5py into the numpy array
+    res = NpyString_pack(
+        info[0].allocator,
+        <npy_packed_static_string*>elem,
+        buf,
+        strlen(buf),
+    )
+    info[0].i += 1
+    return res
+
+
+def vstrings_scatter(hid_t space, size_t contig, size_t noncontig, size_t descr):
+    """Pure-python API interface to _proxy.pyx. This is necessary to
+    allow this module to be conditionally imported, allowing numpy 1.x to
+    continue working everywhere else.
+    """
```

Comment:
This is useful explanation, but I think this more public function should have the longer API docstring.